File: src/processors/savefile_processor.py
import json
from typing import TypedDict
from storage.database_manager import ElementsDatabaseManager


class Element(TypedDict):
    text: str
    emoji: str


def load_json(file_path: str):
    with open(file_path, "r", encoding="utf-8") as f:
        data = json.load(f)
    return data


def process(file_path: str, db_path: str):
    elements_db_manager = ElementsDatabaseManager(db_path)

    data = load_json(file_path)

    elements: list[Element] = data["elements"]

    len_elements = len(elements)

    number_of_batches = 5000
    batch_size = len_elements // number_of_batches

    for i in range(0, len_elements, batch_size):
        print(f"Processing batch {i // batch_size}")
        element_batch = elements[i : i + batch_size]
        try:
            elements_db_manager.bulk_add_elements(
                [(element["text"], element["emoji"]) for element in element_batch]
            )
        except Exception:
            pass

File: src/processors/savefile_processor.ts
import * as z from "zod";

import { ElementsDatabaseManager } from "~/db_manager";

const SavefileSchema = z.object({
  elements: z.array(z.object({ text: z.string(), emoji: z.string() })),
});

export async function process_savefile(file_path: string, db_path: string) {
  const elements_db_manager = new ElementsDatabaseManager(db_path);

  const file = Bun.file(file_path);
  const data = SavefileSchema.parse(await file.json());
  const elements = data.elements;

  const elements_count = elements.length;

  const number_of_batches = 5000;
  const batch_size = Math.floor(elements_count / number_of_batches);

  for (let i = 0; i < elements_count; i += batch_size) {
    console.log(`Processing batch ${Math.round(i / batch_size)}`);
    const element_batch = elements.slice(i, i + batch_size);

    try {
      elements_db_manager.bulk_add_elements(
        element_batch.map((element) => [element.text, element.emoji]),
      );
    } catch {}
  }
}
